### PR TITLE
Add an example on Transformer in Transformer

### PR DESCRIPTION
**Pull request by czy00000**
TNT is consisted of TNT block.Compared to the conventional vision transformers (ViT) which corrupts the local structure
of the patch, the TNT can better preserve and model the local information for visual recognition.

### CODE REVIEW

File: examples/vision/transformer_in_transformer.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    x = tf.image.extract_patches(
+        images=x,
+        sizes=(1, new_patch_size[0], new_patch_size[1], 1),
+        strides=(1, new_patch_size[0], new_patch_size[1], 1),
+        rates=(1, 1, 1, 1),
+        padding="VALID",
+    )
+    x = tf.reshape(x, shape=(-1, in_dim, new_patch_size[0], new_patch_size[1]))
+    # pixel position embedding
+    pixel_position = tf.Variable(
```
Comment (fchollet) [inline examples/vision/transformer_in_transformer.py:109]: Is this supposed to be an element of trainable state? If not it should just be a static constant, not Variable. If yes, then you should be factoring this entire block as a `Layer` subclass.

File: examples/vision/transformer_in_transformer.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+
+"""
+## Implement the TNT block
+"""
+
+
+def transformer_in_transformer_block(
+    pixel_embedding,
+    patch_embedding,
+    out_embedding_dim,
```
Comment (fchollet) [inline examples/vision/transformer_in_transformer.py:138]: Use fully-spelled variable names, e.g. `output_embedding_dim` (only use very standard abbreviations like `dim` or `num`)

File: examples/vision/transformer_in_transformer.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+    in_embedding_dim,
```
Comment (fchollet) [inline examples/vision/transformer_in_transformer.py:139]: Same

File: examples/vision/transformer_in_transformer.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+        patch_embedding, out_embedding_dim * mlp_ratio, out_embedding_dim
+    )
+    patch_embedding = layers.add([patch_embedding, residual_out_2])
+    return pixel_embedding, patch_embedding
+
+
+"""
+## Implement the TNT model
+"""
+
```
Comment (fchollet) [inline examples/vision/transformer_in_transformer.py:188]: Remove break

File: examples/vision/transformer_in_transformer.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+):
+    inputs = layers.Input(shape=input_shape)
+    # Image augment
+    x = data_augmentation(inputs)
+    # extract pixel embedding
+    pixel_embedding, num_patches, new_patch_size = pixel_embed(
+        x, image_size, patch_size, in_embedding_dim, first_stride
+    )
+    num_pixel = new_patch_size[0] * new_patch_size[1]
+    patch_position = tf.Variable(tf.zeros(shape=(1, num_patches, out_embedding_dim)))
```
Comment (fchollet) [inline examples/vision/transformer_in_transformer.py:224]: Same comment

File: examples/vision/transformer_in_transformer.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+"""
+
+loss, accuracy, top_5_accuracy = model.evaluate(x_test, y_test)
+print(f"Test loss: {round(loss, 2)}")
+print(f"Test accuracy: {round(accuracy * 100, 2)}%")
+print(f"Test top 5 accuracy: {round(top_5_accuracy * 100, 2)}%")
+
+"""
+After 30 epochs, the TNT model achieves around 35% accuracy and
+66% top-5 accuracy on the test data. It only has 0.1M parameters.
```
Comment (fchollet) [inline examples/vision/transformer_in_transformer.py:312]: These are rather weak numbers, what's wrong with the setup?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Some preliminary questions/comments for now :)

**Comment by czy00000** [comment]
@fchollet Thank you for your advice. The previous model capacity is too small and may be underfitting. I will increase its capacity slightly and retrain.

**Pull request closed without merging**